### DISCUSSION TITLE
Why isn't the window closing immediately when running a zsh command with `-i` option?

### DISCUSSION THREAD

**Discussion by xfzv** [kovidgoyal/kitty#5148] [Q&A] [answered]
```sh
% kitty --version
kitty 0.25.1 created by Kovid Goyal

% zsh --version
zsh 5.9 (x86_64-pc-linux-gnu)
```
If I run:

```sh
% kitty -e /bin/zsh -i -c "echo hello"
```
a new window appears with "hello" printed, as expected. However, it remains displayed for about 5 seconds before disappearing. Why? with other terminal emulators such as Alacritty or Konsole, the window instantly disappears:

https://user-images.githubusercontent.com/78810647/170892034-269dc31b-3626-4d03-9ee9-0a61fbd7be68.mp4

```sh
% time alacritty -e /bin/zsh -i -c "echo hello"
alacritty -e /bin/zsh -i -c "echo hello"  0,15s user 0,10s system 119% cpu 0,205 total

% time konsole -e /bin/zsh -i -c "echo hello"
konsole -e /bin/zsh -i -c "echo hello"  0,23s user 0,10s system 97% cpu 0,332 total

% time kitty -e /bin/zsh -i -c "echo hello"
kitty -e /bin/zsh -i -c "echo hello"  0,20s user 0,10s system 5% cpu 5,345 total
```

If I run instead

```sh
% kitty -e /bin/bash -i -c "echo hello"
```

then the window disappears instantly (same behavior for every terminal emulator):

```sh
% time alacritty -e /bin/bash -i -c "echo hello"                                                                                
alacritty -e /bin/bash -i -c "echo hello"  0,06s user 0,06s system 79% cpu 0,140 total

% time konsole -e /bin/bash -i -c "echo hello"                                                                                  
konsole -e /bin/bash -i -c "echo hello"  0,18s user 0,06s system 89% cpu 0,270 total

% time kitty -e /bin/bash -i -c "echo hello"
kitty -e /bin/bash -i -c "echo hello"  0,12s user 0,06s system 80% cpu 0,233 total
```

Is there a way to suppress this delay when running a zsh interactive command?

Thanks!

**Comment by kovidgoyal** [comment]
Does not reproduce for me

time kitty -e /bin/zsh -i -c "echo hello"                                                                                                 

real	0.243
user	0.186
sys	0.066
maxmem	119 MB
faults	5


It will be something in your zsh rc files bisect them to find out.

**Reply by xfzv** [reply]
> Does not reproduce for me time kitty -e /bin/zsh -i -c "echo hello"   real 0.243 user 0.186 sys 0.066 maxmem 119 MB faults 5 It will be something in your zsh rc files bisect them to find out.

Thank you. Indeed, turns out the delay is due to [powerlevel10k](https://github.com/romkatv/powerlevel10k). I can reproduce ~consistently~ (Edit: not 100% of the time, sometimes it exits immediately too) without any other zsh related config file, only powerlevel10k.

I wonder why this occurs specifically with kitty and not with Alacritty, Konsole or Urxvt though.

**Comment by kovidgoyal** [comment]
Read the powerlevel10k readme it will tell you

On Mon, May 30, 2022 at 10:32:03AM -0700, xfzv wrote:
> > Does not reproduce for me time kitty -e /bin/zsh -i -c "echo hello"   real 0.243 user 0.186 sys 0.066 maxmem 119 MB faults 5 It will be something in your zsh rc files bisect them to find out.
> 
> Thank you. Indeed, turns out the delay is due to [powerlevel10k](https://github.com/romkatv/powerlevel10k). I can reproduce consistently without any other zsh related config file, only powerlevel10k.
> 
> I wonder why this occurs specifically with kitty and not with Alacritty, Konsole or Urxvt though.
> 
> 
> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/kovidgoyal/kitty/discussions/5148#discussioncomment-2850654
> You are receiving this because you commented.
> 
> Message ID: ***@***.***>

-- 
_____________________________________

Dr. Kovid Goyal 
https://www.kovidgoyal.net
https://calibre-ebook.com
_____________________________________

**Reply by xfzv** [reply, marked as answer]
Fixed with [`close_on_child_death yes`](https://sw.kovidgoyal.net/kitty/conf/?highlight=close_on_child_death#opt-kitty.close_on_child_death), many thanks to @romkatv for [pointing me in the right direction](https://github.com/romkatv/powerlevel10k/issues/1882#issuecomment-1142507727)

**Reply by kovidgoyal** [reply]
That fix would imply that powerlevel10k is forking a background process that keeps the terminal pipe open.

**Reply by romkatv** [reply]
It's possible that it does. There is no harm in that though.
